pkg/test/zq_example.go

```diff
@@ -0,0 +1,224 @@
+package test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/yuin/goldmark"
+	"github.com/yuin/goldmark/ast"
+	"github.com/yuin/goldmark/text"
+)
+
+type ZQExampleBlockType string
+
+const (
+	ZQ_COMMAND ZQExampleBlockType = "zq-command"
+	ZQ_OUTPUT  ZQExampleBlockType = "zq-output"
+)
+
+// Hold a ZQ example as found in markdown.
+type ZQExamplePair struct {
+	command *ast.FencedCodeBlock
+	output  *ast.FencedCodeBlock
+}
+
+// Hold a ZQ example as a testcase found from mardown, derived from a
+// ZQExamplePair.
+type ZQExampleTest struct {
+	Name     string
+	Command  []string
+	Expected string
+}
+
+// Return zq-command / zq-output pairs from a single markdown source after
+// parsing it as a goldmark AST.
+func CollectExamples(node ast.Node, source []byte) ([]ZQExamplePair, error) {
+	var examples []ZQExamplePair
+	var command *ast.FencedCodeBlock
+
+	err := ast.Walk(node, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
+		// Walk() calls its walker func twice. Once when entering and
+		// once before exiting, after walking any children. We need
+		// only do this processing once.
+		if !entering || n == nil || n.Kind() != ast.KindFencedCodeBlock {
+			return ast.WalkContinue, nil
+		}
+
+		fcb, ok := n.(*ast.FencedCodeBlock)
+		if !ok {
+			return ast.WalkStop,
+				fmt.Errorf("likely goldmark bug: Kind() reports a " +
+					"FencedCodeBlock, but the type assertion failed")
+		}
+		bt := ZQExampleBlockType(fcb.Language(source))
+		switch bt {
+		case ZQ_COMMAND:
+			if command != nil {
+				return ast.WalkStop,
+					fmt.Errorf("subsequent %s after another %s", bt, ZQ_COMMAND)
+			}
+			command = fcb
+		case ZQ_OUTPUT:
+			if command == nil {
+				return ast.WalkStop,
+					fmt.Errorf("%s without a preceeding %s", bt, ZQ_COMMAND)
+			}
+			examples = append(examples, ZQExamplePair{command, fcb})
+			command = nil
+			// A fenced code block need not specify an info string, or it
+			// could be arbitrary. The default case is to ignore everything
+			// else.
+		}
+		return ast.WalkContinue, nil
+	})
+
+	if command != nil && err == nil {
+		err = fmt.Errorf("%s without a following %s", ZQ_COMMAND, ZQ_OUTPUT)
+	}
+	return examples, err
+}
+
+// Return the text of a ast.FencedCodeBlock as a string.
+func BlockString(fcb *ast.FencedCodeBlock, source []byte) string {
+	var b strings.Builder
+	for i := 0; i < fcb.Lines().Len(); i++ {
+		line := fcb.Lines().At(i)
+		b.Write(line.Value(source))
+	}
+	return b.String()
+}
+
+// Translate a zq-command example to a runnable command including abspath to zq
+// binary and globs turned into absolute file paths.
+func QualifyCommand(command string) ([]string, error) {
+	command = strings.TrimSpace(command)
+	command = strings.Split(command, "#")[0]
+
+	var delim string
```

Comment:
I suggest we support only one of single or double quotes, and convert the examples as needed, unless there's a requirement for both.

Reply:
Done at 5787f5d including changing the lone example using double quotes.